org.spoofax.interpreter.core/src/main/java/org/spoofax/interpreter/library/ssl/SSL_immutable_relation_filter.java

```diff
@@ -0,0 +1,47 @@
+package org.spoofax.interpreter.library.ssl;
+
+import io.usethesource.capsule.BinaryRelation;
+
+import org.spoofax.interpreter.core.IContext;
+import org.spoofax.interpreter.core.InterpreterException;
+import org.spoofax.interpreter.library.AbstractPrimitive;
+import org.spoofax.interpreter.stratego.Strategy;
+import org.spoofax.interpreter.terms.IStrategoTerm;
+import org.spoofax.interpreter.terms.ITermFactory;
+import org.spoofax.terms.util.TermUtils;
+
+public class SSL_immutable_relation_filter extends AbstractPrimitive {
+
+    protected SSL_immutable_relation_filter() {
+        super("SSL_immutable_relation_filter", 1, 0);
+    }
+
+    @Override
+    public boolean call(IContext env, Strategy[] sargs, IStrategoTerm[] targs) throws InterpreterException {
+        if(!(env.current() instanceof StrategoImmutableRelation)) {
+            return false;
+        }
+        final Strategy mapping = sargs[0];
+        final ITermFactory f = env.getFactory();
+
+        final BinaryRelation.Immutable<IStrategoTerm, IStrategoTerm> relation =
+            ((StrategoImmutableRelation) env.current()).backingRelation;
+        final BinaryRelation.Transient<IStrategoTerm, IStrategoTerm> resultRelation = BinaryRelation.Transient.of();
+        for(java.util.Map.Entry<IStrategoTerm, IStrategoTerm> e : relation.entrySet()) {
+            env.setCurrent(f.makeTuple(e.getKey(), e.getValue()));
+            if(!mapping.evaluate(env)) {
+                continue;
+            }
+            final IStrategoTerm current = env.current();
+            if(!(TermUtils.isTuple(current)) && current.getSubtermCount() == 2) {
+                return false;
```

Comment:
This case is a programming error on the user's side. I prefer to throw an exception in such cases, instead of a silent failure that triggers Stratego's backtracing.